src/executor/stack/executor.rs

```diff
@@ -729,7 +756,15 @@ impl<'config, 'precompiles, S: StackState<'config>, P: PrecompileSet>
 		self.enter_substate(gas_limit, false);
 
 		{
-			if self.code_size(address) != U256::zero() {
+			#[cfg(feature = "with-substrate")]
+			if let Err(e) =
+				self.record_external_operation(crate::ExternalOperation::AddressCodeRead(address))
```

Comment:
Hi, @tgmichel @sorpaas  I am very confused about the `AddressCodeRead` and `IsEmpty` operations here. As far as I know, the `proof_size` submited to the relay chain is used to validate the state transition in the parachain candidate, it's should only contain the storage trie nodes that are necessary to help build the merkle validation tree, right?. Reading evm code storage is only a read operation, no storage changes in this process, why do you need to contains the `AddressCodeRead`  proof_size here?

Reply:
any storage read in a transactional context - like e.g. a evm transaction - will be included in the merkle proof so we need to record its cost.

Reply:
Thanks, are there any docs or discussions that can help me understand this better?  Sounds weird, think about I do a basic contract invocation, change the evm account storage, the final `proof_size` of this dispatch call needs to include the evm code size.